### PR TITLE
Array length constraints

### PR DESCRIPTION
**Pull request by lorchrob**
Defer checking of array length constraints to the model checking phase

### CODE REVIEW

File: src/lustre/lustreNodeGen.ml [hunk 1]
```diff
@@ -1587,6 +1587,28 @@ and compile_node_decl gids is_function cstate ctx i ext inputs outputs locals it
...
+          index_type
+          (Some N.KGhost)
+        in
+        match possible_state_var with
+        | Some state_var -> X.add index state_var accum
+        | None -> accum
+      in let result = X.fold over_indices index_types X.empty in
+      result :: glocals
+    in List.fold_left over_generated_locals glocals gids.GI.array_constraints
+  (* ****************************************************************** *)
```
Comment (daniel-larraz) [inline src/lustre/lustreNodeGen.ml:1611]: Why do you use ad hoc code for these variables? You should add them to the field `locals` in `generatedIdentifiers` at the same time that you add the generated equation.

File: src/lustre/lustreNodeGen.ml [hunk 2]
```diff
@@ -2096,6 +2118,26 @@ and compile_node_decl gids is_function cstate ctx i ext inputs outputs locals it
...
+      let src = Property.Generated (Some pos, [sv]) in
+      a, ac, g, gc, (sv, name, src, Property.Invariant) :: p
+    in
+    let (assumes, _, guarantees, _, props) = 
+      List.fold_left over_array_constraints
+      (assumes, List.length assumes, guarantees, List.length guarantees, props)
+      gids.GI.array_constraints
+    in
+    assumes, guarantees, props
+  (* ****************************************************************** *)
```
Comment (daniel-larraz) [inline src/lustre/lustreNodeGen.ml:2140]: I think only `props` is updated, not `assumes` or `guarantees`. You should simplify the code.

File: tests/regression/falsifiable/arrays_multidimensional3.lus [hunk 3]
```diff
@@ -0,0 +1,11 @@
+node M(const m: int; const n: int) returns (A: [int^m^n, int]);
+let
+  A = {0^2^3, 0};
+tel
+
+node N() returns(A: [int^2^3, int]);
+let
+  A = M(2, 3);
+  check (A = pre A);
```
Comment (daniel-larraz) [inline tests/regression/falsifiable/arrays_multidimensional3.lus:9]: This check is not legal with the latest fix in develop.

Comment (lorchrob) [reply tests/regression/falsifiable/arrays_multidimensional3.lus]: Addressed

File: src/lustre/lustreNodeGen.ml (hunk 2 above)
Comment (lorchrob) [reply src/lustre/lustreNodeGen.ml]: Addressed

File: src/lustre/lustreNodeGen.ml (hunk 1 above)
Comment (lorchrob) [reply src/lustre/lustreNodeGen.ml]: Addressed

**Pull request closed without merging**